### PR TITLE
secrets: read directly from Kubernetes API

### LINKED ISSUE #13012: Tweak Kubernetes secret storage scheme
Our current approach to Kubernetes secret storage is to dump all the secrets into one Kubernetes secret object whose name is provided to `materialized` via the `--user-managed-secret` command-line argument. This has several problems:

  * A single secrets object has a hard size limit of 1MB. This will quickly become a problem for large Materialize Cloud environments.
  * There is a complicated read-modify-write dance to safely update the secret object when adding a new secret without overwriting concurrent writes: https://github.com/MaterializeInc/materialize/blob/0a8ae60853e4d641688df802b8c16cd91ea7661c/src/secrets-kubernetes/src/lib.rs#L112-L144
  * We have to wait for the secrets to propagate through the volume mount before we can consider a secret creation request a success: https://github.com/MaterializeInc/materialize/blob/0a8ae60853e4d641688df802b8c16cd91ea7661c/src/secrets-kubernetes/src/lib.rs#L149-L172. This results in `materialized` modifying secrets mounted upon itself, which is hard to reason about.

@alex-hunt-materialize suggests we instead use the Kubernetes API directly. The secrets controller would create one Kubernetes secret object per secret in Materialize. The secrets reader that @cjubb39 is adding in #12974 would read the secret via an API call upon request.

What we'd lose here is the ability to `ALTER` multiple secrets atomically in the same transaction. That's probably worth it to fix the above limitations.

### PR DESCRIPTION
**Pull request by benesch**
As described in #13012, our current Kubernetes secrets handling scheme
has several deficiencies, most notably that only 1MB of secret data can
be stored in total. This restriction is a result of storing all user
secrets in a single Kubernetes secret.

This commit adjusts our Kubernetes secret integration to instead store
each secret in its own Kubernetes secret. This requires a rather
substantial refactoring of secret plumbing, as we now need to read
secrets by using the Kubernetes API directly, rather than always being
able to read them off the filesystem (via Kubernetes's secret volume
mounts).

Fix #13012.

### Motivation

  * This PR fixes a recognized bug.

### Testing

- [x] This PR has adequate test coverage / QA involvement has been duly considered.

### Release notes

This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):

  - n/a

### CODE REVIEW

File: src/orchestrator-kubernetes/src/secrets.rs [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+                &PatchParams::apply(FIELD_MANAGER).force(),
+                &Patch::Apply(secret),
+            )
+            .await?;
+        Ok(())
+    }
+
+    async fn delete(&self, id: GlobalId) -> Result<(), anyhow::Error> {
+        self.secret_api
+            .delete(&secret_name(id), &DeleteParams::default())
```
Comment (alex-hunt-materialize) [inline src/orchestrator-kubernetes/src/secrets.rs:47]: Heads up that this does not actually wait for it to be deleted by default, nor does it handle cases where the secret already doesn't exist.

You might want to do something like what we do in https://github.com/MaterializeInc/cloud/blob/b476c2d5fe4c6c586c1efa42d0533296807b05ae/mzcloud-shared/src/k8sutil.rs#L66-L103

**Review by alex-hunt-materialize — COMMENTED** [review]
This looks great, except for a couple of details around delete handling.

We'll need to update the environment-controller to not pass the removed arguments to environmentd and to clean up the old secret on apply (temporarily, this can be removed later).

Users will need to manually update any existing secret.

File: src/orchestrator-kubernetes/src/secrets.rs (hunk 1 above)
Comment (benesch) [reply src/orchestrator-kubernetes/src/secrets.rs:57]: Good point about not handling 404. I fixed that. I think we actually _don't_ want to wait for deletion; this happens on the coordinator main thread so better to do the fastest possible thing.

The fact that secrets can leak is a good callout. That was an intentional part of the design, but we never got around to writing the garbage collector. I filed #13579 about that and left a comment in the code. That would handle the cleanup of partially written secret files, too.

**Comment by benesch** [comment]
> and to clean up the old secret on apply (temporarily, this can be removed later).
>
> Users will need to manually update any existing secret.

Can we just nuke all the existing environments? Seems like the easiest thing to do as long as this lands in the next 24 hours.

**Comment by alex-hunt-materialize** [comment]
> Can we just nuke all the existing environments? Seems like the easiest thing to do as long as this lands in the next 24 hours.

That's fine with me.

File: src/orchestrator-kubernetes/src/secrets.rs (hunk 1 above)
Comment (alex-hunt-materialize) [reply src/orchestrator-kubernetes/src/secrets.rs:57]: Works for me.

**Review by alex-hunt-materialize — APPROVED** [review]
You've got some conflicts, but otherwise it looks good.

**Comment by benesch** [comment]
> You've got some conflicts, but otherwise it looks good.

Ah yeah, conflicted with the gRPC PR. Fixed up now. Thanks very much for the review!

**Pull request merged**